Deprecate do_queue kwarg in Operator

frederikwilde opened the pull request:
**Context:**
There are currently two ways to prevent an operator from being queued: the `do_queue` keyword argument and
the `qml.QueuingManager.stop_recording()` context. To remove the former and make the latter the default way
the `do_queue` keyword argument will be deprecated.

**Description of the Change:**
By default the `do_queue` kwarg will be set to `None`, which is operationally equivalent to `do_queue=True`. But
whenever `do_queue` is not `None` a deprecation warning is given to the user. Code using the `do_queue` kwarg
explicitly will still work as before, besides generating deprecation warnings.

**Benefits:**
Users will be informed to switch to `stop_recording`.

**Possible Drawbacks:**
In `QuantumScript` currently `do_queue=False` is set as a class variable. This might lead to unexpected warnings,
even when the user has never specified `do_queue` explicitly. This also needs to be addressed, when `do_queue` gets
removed.

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/161

frederikwilde commented:
[SC-37934]

albi3ro commented on tests/test_operation.py:
```diff
@@ -87,7 +97,8 @@ class DummyOp(qml.operation.Operator):
 
         with pytest.raises(qml.wires.WireError, match="Wires must be unique"):
-            DummyOp(0.5, wires=[1, 1], do_queue=False)
+            with qml.QueuingManager.stop_recording():
```
I wonder why `do_queue=False` was used here at all.  It's testing something completely separate. We can probably just get rid of the keyword argument entirely. Shouldn't change the content of test.

frederikwilde replied:
You are right. It's not necessary. I removed it 0b784cf1b1bd4110717c5a45410c6c7b9baec454

albi3ro commented on tests/ops/functions/test_eigvals.py:
```diff
@@ -327,6 +327,21 @@ def test_composite_eigvals(self):
 
         assert np.allclose(mat_eigvals, op_eigvals)
 
+    @pytest.mark.parametrize("do_queue", [True, False])
+    def test_composite_do_queue_deprecation(self, do_queue):
+        """Test that CompositeOp raises a deprecation warning when do_queue is not None."""
+        factors = (qml.PauliX(0), qml.PauliY(0))
+        do_queue_deprecation_warning = (
+            "The do_queue keyword argument is deprecated. "
+            "Use qml.queuing.QueuingManager.stop_recording()"
+        )
+
+        with pytest.warns(UserWarning, match=do_queue_deprecation_warning):
+            qml.prod(*factors, do_queue=do_queue)
+
+        with pytest.warns(UserWarning, match=do_queue_deprecation_warning):
+            qml.sum(*factors, do_queue=do_queue)
+
```
I think this change may be to the wrong test file.  This looks to be in `test_eigvals.py`. It should probably be in `ops/op_math/test_composite.py`.

frederikwilde commented:
There is one behaviour which is not ideal at the moment and will also have to be figured out when `do_queue` gets removed. `QuantumScript` has a class variable `do_queue` which is `False` by default, which will now lead to a deprecation warning, for instance when calling `.adjoint()`.

When instantiating a `QuantumTape` this variable will be overwritten, by default to `None`, which is equivalent to `True`, but suppresses the warning.

albi3ro commented on doc/releases/changelog-dev.md:
```diff
@@ -143,6 +143,9 @@
 * `qml.grouping` module is removed. The functionality has been reorganized in the `qml.pauli` module.
 
+* `do_queue` keyword argument in `qml.operation.Operator` is deprecated. Please
+  use the `qml.QueuingManager.stop_recording()` context instead.
```
```suggestion
  use the `qml.QueuingManager.stop_recording()` context instead.
  [(#4148)](https://github.com/PennyLaneAI/pennylane/pull/4148)
```

albi3ro commented on pennylane/operation.py:
```diff
@@ -420,8 +420,9 @@ class Operator(abc.ABC):
-        do_queue (bool): indicates whether the operator should be
-            recorded when created in a tape context
+        do_queue (bool): indicates whether the operator should be recorded when created in
+            a tape context. This arguement is deprecated, instead of setting it to `False`
+            use `qml.QueuingManager.stop_recording()`.
```
```suggestion
            a tape context. This arguement is deprecated, instead of setting it to ``False``
            use :meth:`~.QueuingManager.stop_recording`.
```
rst and Sphinx are annoying. Code like `False` needs double back ticks.  If we do something like `:meth:~.QueuingManager.stop_recording`, sphinx will insert a link to the relevant method. I think I got that syntax right at least. I might have done something wrong.


![Screenshot 2023-05-19 at 2 35 50 PM](https://github.com/PennyLaneAI/pennylane/assets/6364575/0221d640-a06e-412b-a2de-e4cce385b8f0)

frederikwilde replied:
Oh, dear. Thanks. I changed all docstrings now: ef7e75f7809a8cbcb168a647c4d99002c90df60d

albi3ro reviewed (commented):
Preliminary review so you get comment about rst formatting. The comment will need to be duplicated across all docstrings.

albi3ro reviewed (approved):
87 files changed!

We should get a final sign off on the length of the deprecation cycle, but I'm happy to approve these changes :)

dwierichs commented on pennylane/operation.py:
```diff
@@ -420,8 +420,9 @@ class Operator(abc.ABC):
-        do_queue (bool): indicates whether the operator should be
-            recorded when created in a tape context
+        do_queue (bool): indicates whether the operator should be recorded when created in
+            a tape context. This arguement is deprecated, instead of setting it to ``False``
+            use :meth:`~.QueuingManager.stop_recording`.
```
I really like the explicit statement of which action before the deprecation should be replaced by which action after the deprecation! Maybe we could be as explicit in the changelog and the deprecations doc page? :)

frederikwilde replied:
Good point. Done. 94938879d06f40ec63a5a2078308be04e250fe55

dwierichs commented on pennylane/operation.py:
```diff
@@ -1023,7 +1026,14 @@ def __init__(self, *params, wires=None, do_queue=True, id=None):
-        if do_queue:
+        if do_queue is not None:
+            do_queue_deprecation_warning = (
+                "The do_queue keyword argument is deprecated. "
+                "Use qml.queuing.QueuingManager.stop_recording()"
```
```suggestion
                "The do_queue keyword argument is deprecated. Instead of setting "
                "it to false, use qml.queuing.QueuingManager.stop_recording()"
```

dwierichs commented on pennylane/ops/identity.py:
```diff
@@ -46,7 +46,7 @@ class Identity(CVObservable, Operation):
     ev_order = 1
 
-    def __init__(self, *params, wires=None, do_queue=True, id=None):
+    def __init__(self, *params, wires=None, do_queue=None, id=None):
```
Not particularly related to this PR, but should the docstring have an "Args: ... " section? :thinking:

frederikwilde replied:
Filled in the missing arguments in the docstrings: 94915cd8a1748832d3d6d4ef3b35423e7b8ecb0b
I don't think we need to have anything on `*params`, since it doesn't have any effect, right?

dwierichs replied:
Thanks for the extra mile! :)
No, `params` doesn't need a comment I think :)

dwierichs commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -151,7 +151,7 @@ class QubitStateVector(StatePrep):
     """int: Number of dimensions per trainable parameter of the operator."""
 
-    def __init__(self, state, wires, do_queue=True, id=None):
+    def __init__(self, state, wires, do_queue=None, id=None):
```
Could add `do_queue` and `id` to the "Args" section above.

frederikwilde replied:
94915cd8a1748832d3d6d4ef3b35423e7b8ecb0b

dwierichs commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -60,7 +60,7 @@ class QutritUnitary(Operation):
     """Gradient computation method."""
 
-    def __init__(self, *params, wires, do_queue=True):
+    def __init__(self, *params, wires, do_queue=None):
```
Could add `do_queue` and `id` to the "Args" section above.

dwierichs commented on pennylane/ops/snapshot.py:
```diff
@@ -61,7 +61,7 @@ def circuit():
     grad_method = None
 
-    def __init__(self, tag=None, do_queue=True):
+    def __init__(self, tag=None, do_queue=None):
```
Could add `do_queue` and `id` to the "Args" section above.

frederikwilde replied:
94915cd8a1748832d3d6d4ef3b35423e7b8ecb0b

dwierichs commented on pennylane/queuing.py:
```diff
@@ -151,11 +151,10 @@ def qfunc(x, scale_value=1):
-In order to construct new operators within a recording, but without queuing them, either
-use the :meth:`~.QueuingManager.stop_recording` context or specify `do_queue=False` upon construction:
+In order to construct new operators within a recording, but without queuing them
+use the :meth:`~.QueuingManager.stop_recording` contextupon construction:
```
```suggestion
use the :meth:`~.QueuingManager.stop_recording` context upon construction:
```

dwierichs commented on pennylane/tape/operation_recorder.py:
```diff
@@ -50,7 +50,6 @@ def __init__(
         prep=None,
         shots=None,
         name=None,
-        do_queue=False,
```
I'm not familiar with this recorder, are we fine just removing this kwarg?

frederikwilde replied:
Hm, I don't know either. The argument is ignored by `OperationRecorder` and the class has no children. Also, whenever `OperationRecorder` is instantiated in PennyLane, no arguments are passed. I'd assume if all tests pass with this change, it should be fine.

dwierichs commented on pennylane/tape/qscript.py:
```diff
@@ -1130,7 +1130,15 @@ def adjoint(self):
         )
-        if self.do_queue:
+
+        if self.do_queue is not None:
+            do_queue_deprecation_warning = (
+                "The do_queue keyword argument is deprecated. "
+                "Use qml.queuing.QueuingManager.stop_recording()"
+            )
+            warnings.warn(do_queue_deprecation_warning, UserWarning)
```
Reading this repeatedly gives me ideas of having a central `deprecations.py` file with e.g. warnings. This might make adding and removing deprecations more handy :) 
Just a side note, though :)

frederikwilde replied:
Absolutely! In such a file we could also manage how often warnings are issued, e.g., using
```Python
warnings.filterwarnings("once", "The do_queue keyword argument is deprecated.", UserWarning)
```
In this PR I tried to make sure that the deprecation warning is not issued more than once, but sometimes that is impossible to achieve manually. @trbromley and I also talked about this.

dwierichs replied:
Ah, that's a nice feature of this as well :) 
Not time sensitive but definitely would be nice to have!

dwierichs replied:
(I'll resolve, because this topic is closed for this PR)

dwierichs commented on pennylane/tape/tape.py:
```diff
@@ -327,13 +330,14 @@ class QuantumTape(QuantumScript, AnnotatedQueue):
 
     .. code-block:: python
 
         with qml.tape.QuantumTape() as tape1:
-            with qml.tape.QuantumTape(do_queue=False) as tape2:
-                qml.RX(0.123, wires=0)
+            with qml.QueuingManager.stop_recording():
+                with qml.tape.QuantumTape() as tape2:
+                    qml.RX(0.123, wires=0)
```
:100:

dwierichs commented on pennylane/templates/embeddings/amplitude.py:
```diff
@@ -121,7 +121,7 @@ def circuit(f=None):
     grad_method = None
 
-    def __init__(self, features, wires, pad_with=None, normalize=False, do_queue=True, id=None):
+    def __init__(self, features, wires, pad_with=None, normalize=False, do_queue=None, id=None):
```
Could add `do_queue` and `id` to the "Args" section above.

dwierichs commented on pennylane/templates/embeddings/angle.py:
```diff
@@ -73,7 +73,7 @@ def circuit(feature_vector):
     grad_method = None
 
-    def __init__(self, features, wires, rotation="X", do_queue=True, id=None):
+    def __init__(self, features, wires, rotation="X", do_queue=None, id=None):
```
Could add `do_queue` and `id` to the "Args" section above.

frederikwilde replied:
94915cd8a1748832d3d6d4ef3b35423e7b8ecb0b

dwierichs commented on pennylane/templates/subroutines/hilbert_schmidt.py:
```diff
@@ -71,8 +71,9 @@ class HilbertSchmidt(Operation):
 
-            with qml.tape.QuantumTape(do_queue=False) as u_tape:
-                qml.Hadamard(wires=0)
+            with qml.tape.QuantumTape() as u_tape:
+                with qml.QueuingManager.stop_recording():
```
I think we need to flip the order, no?
```suggestion
            with qml.QueuingManager.stop_recording():
                with qml.tape.QuantumTape() as u_tape:
```
This way, `Hadamard` will be in `u_tape`, but `u_tape` will not be queued in an outer context.

frederikwilde replied:
Oh, nice catch!

dwierichs commented on pennylane/templates/subroutines/hilbert_schmidt.py:
```diff
@@ -213,8 +214,9 @@ class LocalHilbertSchmidt(HilbertSchmidt):
 
-            with qml.tape.QuantumTape(do_queue=False) as u_tape:
-                qml.CZ(wires=[0,1])
+            with qml.tape.QuantumTape() as u_tape:
+                with qml.QueuingManager.stop_recording():
```
Same as above
```suggestion
            with qml.QueuingManager.stop_recording():
                with qml.tape.QuantumTape() as u_tape:
```

dwierichs commented on tests/ops/op_math/test_sum.py:
```diff
@@ -85,7 +85,7 @@ def _get_pw(w, pauli_op):
 
 # pylint: disable=unused-argument
-def sum_using_dunder_method(*summands, do_queue=False, id=None):
+def sum_using_dunder_method(*summands, id=None):
```
I'm not entirely sure, but aren't you calling this function with `do_queue=None` further below? :thinking:

frederikwilde replied:
Oh nice, catch. I hadn't gone through all the tests yet. Removed the argument in the function call: 20d4ab5cda0501ccefc193e90e263d62f0ff55bd

dwierichs reviewed (commented):
Wow, very thorough work @frederikwilde :100: Thanks a lot for this, so many files!

I found the following lines we still may want to change (can't suggest this in Github I think):
- [x] `pulse/parametrized_evolution.py::line42` 
- [x] `tape/tape.py::line247` (Might as well remove the manual defaults from the docstring)
- [x] `tests/test_operation.py::TestOperatorConstruction::test_operation_outside_context` Either remove the different options of `do_queue` or put a `with pytest.warns(...)` context around the respective tests :) 
- [x] `tests/test_operation.py::line2532`
- [x] `tests/ops/op_math/test_prod.py::line1198` not really clear to me why that line is there at all.
- [ ] `tests/ops/op_math/test_sprod.py::line854` same
- [ ] `tests/ops/op_math/test_sum.py::line917` same
- [ ] `tests/ops/op_math/test_adjoint.py::line553` wrap this line in a `pytest.warns` context to catch the warning
- [ ] `tests/ops/op_math/test_controlled.py::line636` same
- [ ] `tests/ops/op_math/test_pow_op.py::line620` same
- [ ] `tests/ops/op_math/test_symbolic_op.py::line202` same
- [ ] `tests/templates/test_subroutines/test_grover.py::line51` same
- [ ] `tests/ops/op_math/test_prod.py::line356` I think it's ok to just remove the explicit `do_queue` setting.

frederikwilde commented:
Thanks @dwierichs and @albi3ro for the thorough review! I addressed all comments.
@dwierichs, I left some comments unresolved. Feel free to take another look.

dwierichs commented on pennylane/tape/tape.py:
```diff
@@ -355,16 +359,22 @@ def __init__(
+        if do_queue is not None:
+            do_queue_deprecation_warning = (
+                "The do_queue keyword argument is deprecated. "
+                "Use qml.queuing.QueuingManager.stop_recording()"
```
```suggestion
                "Instead of setting it to False, use qml.queuing.QueuingManager.stop_recording()"
```

dwierichs commented on pennylane/ops/op_math/symbolicop.py:
```diff
@@ -66,13 +69,20 @@ def __copy__(self):
+        if do_queue is not None:
+            do_queue_deprecation_warning = (
+                "The do_queue keyword argument is deprecated. "
+                "Use qml.queuing.QueuingManager.stop_recording()"
```
```suggestion
                "Instead of setting it to False, use qml.queuing.QueuingManager.stop_recording()"
```

dwierichs commented on pennylane/tape/qscript.py:
```diff
@@ -1152,7 +1152,15 @@ def adjoint(self):
+        if self.do_queue is not None:
+            do_queue_deprecation_warning = (
+                "The do_queue keyword argument is deprecated. "
+                "Use qml.queuing.QueuingManager.stop_recording()"
```
```suggestion
                "Instead of setting it to False, use qml.queuing.QueuingManager.stop_recording()"
```

dwierichs reviewed (approved):
Thanks @frederikwilde! Great job :) 
Just suggested a few changes to match between different warning texts / the test regex.
There seems to be some problem with the linking to `stop_recording` in the documentation, not quite sure what's the cause of this?
Good to go from my side once these two things are resolved :)

albi3ro commented on pennylane/ops/identity.py:
```diff
@@ -32,6 +32,14 @@ class Identity(CVObservable, Operation):
     .. math::
         E[\I] = \text{Tr}(\I \rho)
 
+    Kwargs:
+        wires (Iterable[Any] or Any): Wire label(s) that the identity acts on.
+        do_queue (bool): indicates whether the operator should be recorded when created in
+            a tape context. This argument is deprecated, instead of setting it to ``False``
+            use :meth:`~.queuing.QueuingManager.stop_recording`.
+        id (str): custom label given to an operator instance,
+            can be useful for some applications where the instance has to be identified.
+
     corresponds to the trace of the quantum state, which in exact
     simulators should always be equal to 1.
     """
```
```suggestion

    corresponds to the trace of the quantum state, which in exact
    simulators should always be equal to 1.

    Args:
        wires (Iterable[Any] or Any): Wire label(s) that the identity acts on.
        do_queue (bool): indicates whether the operator should be recorded when created in
            a tape context. This argument is deprecated, instead of setting it to ``False``
            use :meth:`~.queuing.QueuingManager.stop_recording`.
        id (str): custom label given to an operator instance,
            can be useful for some applications where the instance has to be identified.
    """
```

something to try.
